datafusion/optimizer/src/replace_distinct_aggregate.rs

```diff
@@ -88,60 +94,72 @@ impl OptimizerRule for ReplaceDistinctWithAggregate {
                 input,
                 schema,
             })) => {
+                let expr_cnt = on_expr.len();
+
                 // Construct the aggregation expression to be used to fetch the selected expressions.
-                let aggr_expr = select_expr
-                    .iter()
-                    .map(|e| {
-                        Expr::AggregateFunction(AggregateFunction::new(
-                            AggregateFunctionFunc::FirstValue,
-                            vec![e.clone()],
-                            false,
-                            None,
-                            sort_expr.clone(),
-                            None,
-                        ))
-                    })
-                    .collect::<Vec<Expr>>();
+                let aggr_expr = vec![Expr::AggregateFunction(AggregateFunction::new(
+                    AggregateFunctionFunc::FirstValue,
+                    select_expr,
+                    false,
+                    None,
+                    sort_expr.clone(),
+                    None,
+                ))];
```

Comment:
I don't know if they are semantically identical.
Instructions welcomed.

Reply:
CI failed, i guess the answer is no. :(

Reply:
The difference is that the `select_expr.iter()...` makes a `vec` with the same number of elements as `select_expr `

To avoid a clone, perhaps you can use `into_iter` instead of `iter`, like

```rust
                let aggr_expr = select_expr
                    .into_iter() // <---- Use into_iter() here
                    .map(|e| {
                        Expr::AggregateFunction(AggregateFunction::new(
                            AggregateFunctionFunc::FirstValue,
                            vec![e],    // <--- to avoid a clone here
                            false,
                            None,
                            sort_expr.clone(),
                            None,
                        ))
                    })
                    .collect::<Vec<Expr>>();
```

Reply:
> The difference is that the `select_expr.iter()...` makes a `vec` with the same number of elements as `select_expr `
> 
> To avoid a clone, perhaps you can use `into_iter` instead of `iter`, like
> 
> ```rust
>                 let aggr_expr = select_expr
>                     .into_iter() // <---- Use into_iter() here
>                     .map(|e| {
>                         Expr::AggregateFunction(AggregateFunction::new(
>                             AggregateFunctionFunc::FirstValue,
>                             vec![e],    // <--- to avoid a clone here
>                             false,
>                             None,
>                             sort_expr.clone(),
>                             None,
>                         ))
>                     })
>                     .collect::<Vec<Expr>>();
> ```

What bothers me most is that I cannot avoid cloning `sort_expr`. However, it looks like the only way to make it correct.